imapclient/exceptions.py

```diff
@@ -0,0 +1,35 @@
+"""
+Global IMAPClient exceptions
+"""
+import imaplib
+
+
+# Base class allowing to catch any IMAPClient related exceptions
+# To ensure backward compatibility, we "rename" the imaplib general
+# exception class, so we can catch its exceptions without having to
+# deal with it in IMAPClient codebase
+IMAPClientException = imaplib.IMAP4.error
+
+
+class CapabilityError(IMAPClientException):
+    """ A command that needs a capability not installed on the server """
+
+
+class LoginFailed(IMAPClientException):
+    """ An error occurred during the IMAP login """
+
+
+class NoSelectedFolder(IMAPClientException):
+    """ A folder needs to be selected before executing the command """
+
+
+class ReadOnlyError(IMAPClientException):
+    """ Raised if a modifying operation was attempted on a read-only folder """
+
+
+class InvalidSearchCriteria(IMAPClientException):
+    """ The SEARCH command failed, probably due to wrong criteria """
```

Comment:
Perhaps generalise this to `InvalidCriteriaError` so that the name work better for `sort()` and `thread()` criteria too?